beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -521,20 +521,119 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+    ///
+    /// ## Notes
+    ///
+    /// - Use the `skips` parameter to define the behaviour when `request_slot` is a skipped slot.
+    /// - Returns `Ok(None)` for any slot higher than the current wall-clock slot.
+    pub fn block_root_at_slot(
+        &self,
+        request_slot: Slot,
+        skips: WhenSlotSkipped,
+    ) -> Result<Option<Hash256>, Error> {
+        match skips {
+            WhenSlotSkipped::None => self.block_root_at_slot_skips_none(request_slot),
+            WhenSlotSkipped::Prev => self.block_root_at_slot_skips_prev(request_slot),
+        }
+    }
+
+    /// Returns the block root at the given slot, if any. Only returns roots in the canonical chain.
+    ///
+    /// ## Notes
+    ///
+    /// - Returns `Ok(None)` if the given `Slot` was skipped.
+    /// - Returns `Ok(None)` for any slot higher than the current wall-clock slot.
+    ///
+    /// ## Errors
+    ///
+    /// May return a database error.
+    fn block_root_at_slot_skips_none(&self, request_slot: Slot) -> Result<Option<Hash256>, Error> {
+        if request_slot > self.slot()? {
+            return Ok(None);
+        } else if request_slot == self.spec.genesis_slot {
+            return Ok(Some(self.genesis_block_root));
+        }
+
+        let prev_slot = request_slot.saturating_sub(1_u64);
+
+        // Try an optimized path of reading the root directly from the head state.
+        let fast_lookup: Option<Option<Hash256>> = self.with_head(|head| {
+            let state = &head.beacon_state;
+
+            // Try find the root for the `request_slot`.
+            let request_root_opt = match state.slot.cmp(&request_slot) {
+                // It's always a skip slot if the head is less than the request slot, return early.
+                Ordering::Less => return Ok(Some(None)),
+                // The request slot is the head slot.
+                Ordering::Equal => Some(head.beacon_block_root),
+                // Try find the request slot in the state.
+                Ordering::Greater => state.get_block_root(request_slot).ok().copied(),
+            };
+
+            if let Some(request_root) = request_root_opt {
+                if let Ok(prev_root) = state.get_block_root(prev_slot) {
+                    return Ok(Some((*prev_root != request_root).then(|| request_root)));
+                }
+            }
+
+            // Fast lookup is not possible.
+            Ok::<_, Error>(None)
+        })?;
+        if let Some(root_opt) = fast_lookup {
+            return Ok(root_opt);
+        }
+
+        let mut prev_root_opt = None;
+        process_results(self.forwards_iter_block_roots(prev_slot)?, |iter| {
+            for (curr_root, curr_slot) in iter {
+                if let Some(prev_root) = prev_root_opt {
+                    if curr_slot == request_slot {
+                        return (curr_root != prev_root).then(|| curr_root);
+                    }
+                }
+                prev_root_opt = Some(curr_root);
```

Comment:
This is a minor style thing, but I think this loop could be more neatly written with `tuple_windows` like:

```rust
for ((prev_root, _), (curr_root, curr_slot)) in iter.tuple_windows() {
    if curr_slot == request_slot {
        return (curr_root != prev_root).then(|| curr_root);
    }
}
```

Further, seeing as we know that the request should be the first item of the `tuple_windows` iterator, we could use `next()`, with an additional check on the slot:

```rust
if let Some((prev_root, _), (curr_root, curr_slot)) = process_results(self.forwards_iter_block_roots(prev_slot)?, |iter| {
    iter.tuple_windows().next()
}) {
    if curr_slot != request_slot {
        return Err(..);
    }
    return Ok((curr_root != prev_root).then(|| curr_root));
} else {
    return Ok(None);
}
```

Idk though, that's a bit messier.